### PR TITLE
make ReadRepository more robust

### PR DESCRIPTION
**Pull request by listx**
This addresses the concerns in #75. There are 2 key changes:

(1) we do *not* proceed to call `extractRegistryTags` if the network call errored out for some reason. This alone should avoid the panics we have seen as described in #75.

(2) we use retry logic with backoff from k8s.io/apimachinery. This should increase the robustness of ReadRepository when Prow (where CIP runs) gets throttled by GCR.

This isn't a complete fix, because we should invalidate those parts of the registry inventory state that we have trouble populating; these sloppy areas should be removed from the promotion set difference logic entirely. That item is captured as a TODO item, and is the next thing to tackle. I wanted to put this up for review though, to get some feedback before I write too much code.

### CODE REVIEW

**Comment by dims** [comment]
/assign @amy 

i took a quick look, the changes looks reasonable to me :)

File: lib/dockerregistry/inventory.go [hunk 1]
```diff
@@ -373,29 +374,57 @@ func (riid *RegInvImageDigest) PrettyValue() string {
...
+		googleTags, err = getRegistryTagsFrom(req)
+
+		// We never return an error (err) in the second part of our return
+		// argument, because we don't want to prematurely stop the
+		// ExponentialBackoff() loop; we want it to continue looping until
+		// either we get a well-formed tags value, or until it hits
+		// ErrWaitTimeout. This is how ExponentialBackoff() uses the
+		// ConditionFunc type.
+		if err == nil && googleTags != nil {
+			return true, nil
```
Comment (justinsb) [inline lib/dockerregistry/inventory.go]: Maybe log the error though?

File: lib/dockerregistry/inventory.go [hunk 2]
```diff
@@ -373,29 +374,57 @@ func (riid *RegInvImageDigest) PrettyValue() string {
...
-		return nil, errors
+		return nil, err
 	}
 
 	tags, err := extractRegistryTags(reader)
 	if err != nil {
-		errors = append(errors, Error{
-			Context: "parsing JSON",
-			Error:   err})
+		klog.Warning("error parsing *google.Tags from io.Reader handle:", err)
```
Comment (justinsb) [inline lib/dockerregistry/inventory.go]: Probably `klog.Warningf("error parsing tags: %v", err)` ; I don't think the types are important enough to be in the error message.

File: lib/dockerregistry/inventory.go [hunk 3]
```diff
@@ -373,29 +374,57 @@ func (riid *RegInvImageDigest) PrettyValue() string {
...
+		return nil, err
+	}
+
+	return googleTags, nil
+}
+
+func getRegistryTagsFrom(req stream.ExternalRequest) (*google.Tags, error) {
 	defer req.StreamProducer.Close()
 
-	errors := make(Errors, 0)
```
Comment (justinsb) [inline lib/dockerregistry/inventory.go]: Nice change - makes the error handling much clearer :+1:

**Comment by justinsb** [comment]
Looks good to me!

/lgtm

**Comment by listx** [comment]
Oops, should probably have put a `/hold` to let Amy chime in as well. I will create a followup PR to address all comments from this PR, so please still take a look if you haven't already.

**Pull request merged**